src/lib/jobs.rs

```diff
@@ -0,0 +1,346 @@
+            .ok_or(RLWServerError(
+                "There is no running process associated with this job".to_string(),
+            ))?
+            .to_string();
+        mem::drop(lock_handle);
+
+        // Send kill signal
+        let thread = thread::spawn(move || -> Result<ExitStatus, RLWServerError> {
+            if forced {
+                return execute_command(
+                    "kill".to_string(),
+                    vec!["-9".to_string(), pid],
+                    None,
+                    &tx_output,
+                );
+            }
+            return execute_command(
+                "kill".to_string(),
+                vec!["-15".to_string(), pid],
+                None,
+                &tx_output,
+            );
+        });
+
+        // Populate stdout/stderr output
+        for rec in rx_output {
+            self.output.lock().await.push(rec);
+        }
+
+        thread.join().map_err(|e| {
+            RLWServerError(format!("Error joining on kill signal thread {:?}", e))
+        })??;
+
+        Ok(())
+    }
+
+    /// Stream the history and all upcoming output from `self.start_command()`.
+    ///
+    /// # Arguments
+    /// * `self: Arc<Job>` - Arc of self. Required in order to use self.output in an async tokio thread.
+    ///
+    /// # Returns
+    ///
+    /// The function returns a result with a tuple containing the following types:
+    /// * StreamResponse Receiver - type to be used by tonic to stream the output to the client.
+    /// * JoinHandle              - used to propagate errors up to the main thread in order to be handled.
+    pub async fn stream_job(
+        self: Arc<Job>,
+    ) -> Result<
+        (
+            tokio_mpsc::Receiver<Result<StreamResponse, Status>>,
+            JoinHandle<Result<(), RLWServerError>>,
+        ),
+        RLWServerError,
+    > {
+        let (tx, rx) = tokio_mpsc::channel(STREAM_BUFFER_SIZE);
+        let stream_handle = tokio::spawn(async move {
+            // An index into job.output is used to determine if any new values need
+            // to be streamed to the client.
+            let mut read_idx;
+
+            // Send the client the job history
+            {
+                let output_guard = self.output.lock().await;
+                let res = tx
+                    .send(Ok(StreamResponse {
+                        output: output_guard.clone(),
+                    }))
+                    .await;
+
+                if let Err(e) = res {
+                    return Err(RLWServerError(format!(
+                        "Error sending job history: {:?}",
+                        e
+                    )));
+                }
+                read_idx = output_guard.len();
+            }
+
+            // While the job is running, send new output messages
+            while matches!(
+                *self.status.lock().await,
+                status_response::ProcessStatus::Running(true)
+            ) {
```

Comment:
This is pretty unusual. What about using `while let ...`?

```suggestion
            while let status_response::ProcessStatus::Running(true) = *self.status.lock().await {
```

Reply:
I was unaware that the if let ... syntax extended to while lets also, I like it thanks.

Reply:
Resolved.